packages/headless/src/features/insight-user-actions/insight-user-actions-preprocessing.ts

```diff
@@ -0,0 +1,339 @@
+import {
+  UserAction,
+  UserActionsState,
+  UserActionTimeline,
+  UserActionType,
+  UserSession,
+} from './insight-user-actions-state';
+
+const SECOND = 1000;
+const MINUTE = SECOND * 60;
+const SESSION_INACTIVITY_THRESHOLD_IN_MS = 30 * MINUTE;
+const NUMBER_OF_PRECEDING_SESSIONS = 2;
+const NUMBER_OF_FOLLOWING_SESSIONS = 2;
+
+export interface RawUserAction {
+  name: string;
+  value: string;
+  time: string;
+}
+
+/**
+ * Preprocesses the raw user actions data from the API into a user actions timeline with the following steps:
+ * - The data is first sorted by most recent and mapped into user actions.
+ * - The actions are then split into sessions based on the ticket creation date.
+ * - The sessions are then stored in the timeline object either as preceding, following or the case creation session.
+ * - The timeline is then filtered to exclude custom actions passed in the state `excludedCustomActions` and empty search queries.
+ * Note: The filtering is done at the end to avoid filtering out actions that would lead to incorrectly split a session into two by creating a gap.
+ * @param state {UserActionsState} - The state of the user actions
+ * @param actions {rawUserAction[]} - The raw user actions array to preprocess
+ * @returns {UserActionTimeline}
+ */
+export const preprocessActionsData = (
+  state: UserActionsState,
+  actions: Array<RawUserAction>
+): UserActionTimeline => {
+  if (!state.ticketCreationDate || actions.length === 0) {
```

Comment:
Also security needed here, in the case where the API doesn't return an actions array.
```suggestion
  if (!state.ticketCreationDate || actions?.length === 0) {
```